### PR TITLE
[BEAM-3489] add PubSub messageId in PubsubMessage

### PR DESCRIPTION
**Pull request by thinhha**
Make the `recordId` field from `PubsubClient.IncomingMessage` available as a new field called `messageId` in `PubsubMessage`.
Updated the coder for `PubsubMessage` to encode and decode the `messageId`.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by thinhha** [comment]
R: @lukecwik

**Comment by thinhha** [comment]
retest this please

**Comment by thinhha** [comment]
Thanks - I've added a few changes based on your comment. It's my first PR so sorry for all the questions!

* For PubsubMesageWithMessageIdCoder, do I need to explicitly skip the attributes map which is second in pubsub.proto before getting to the message_id (which is third)?
* Where would be a good place to add some tests for this PR?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessage.java [hunk 1]
```diff
@@ -23,17 +23,25 @@
...
 public class PubsubMessage {
 
   private byte[] message;
   private Map<String, String> attributes;
+  private String messageId;
 
   public PubsubMessage(byte[] payload, Map<String, String> attributes) {
     this.message = payload;
     this.attributes = attributes;
+    this.messageId = "";
```
Comment (lukecwik) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessage.java]: I think you should set this to null.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessageWithAttributesAndMessageIdCoder.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.Map;
+import org.apache.beam.sdk.coders.ByteArrayCoder;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CustomCoder;
+import org.apache.beam.sdk.coders.MapCoder;
+import org.apache.beam.sdk.coders.NullableCoder;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.values.TypeDescriptor;
+
+/** A coder for PubsubMessage including attributes. */
```
Comment (lukecwik) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessageWithAttributesAndMessageIdCoder.java]: please update comment to state that you also include the message id.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessageWithMessageIdCoder.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import org.apache.beam.sdk.coders.ByteArrayCoder;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CustomCoder;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.collect.ImmutableMap;
+
+/** A coder for PubsubMessage treating the raw bytes being decoded as the message's payload. */
```
Comment (lukecwik) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessageWithMessageIdCoder.java]: Please update this comment as well that it contains the message id.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessage.java [hunk 4]
```diff
@@ -23,17 +23,25 @@
 import javax.annotation.Nullable;
 
 /**
- * Class representing a Pub/Sub message. Each message contains a single message payload and a map of
- * attached attributes.
+ * Class representing a Pub/Sub message. Each message contains a single message payload, a map of
+ * attached attributes, and an optional messageId.
```
Comment (lukecwik) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessage.java]:
```suggestion
 * attached attributes, and a message id.
```

**Comment by lukecwik** [comment]
Looking good so far, minor comment updates and some tests and this will be merged.

**Comment by lukecwik** [comment]
@thinhha any updates?

**Comment by thinhha** [comment]
Sorry for the delay! I've added some changes based on your comment.

I've added an additional integration test, hopefully it will be OK.

**Comment by lukecwik** [comment]
Run Java PostCommit

**Comment by lukecwik** [comment]
I kicked off the postcommit to validate that the IT you added.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubReadIT.java [hunk 5]
```diff
@@ -61,4 +62,35 @@ public void testReadPublicData() throws Exception {
...
+        pipeline.apply(
+            PubsubIO.readMessagesWithMessageId()
+                .fromTopic("projects/pubsub-public-data/topics/taxirides-realtime"));
+
+    messages.apply(
+        "isMessageIdNonNull",
+        signal.signalSuccessWhen(
+            messages.getCoder(),
+            pubsubMessages ->
+                pubsubMessages.stream().noneMatch(m -> Strings.isNullOrEmpty(m.getMessageId()))));
```
Comment (lukecwik) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubReadIT.java]: It looks like the new IT doesn't pass:
https://builds.apache.org/job/beam_PostCommit_Java_PR/169/testReport/junit/org.apache.beam.sdk.io.gcp.pubsub/PubsubReadIT/testReadPubsubMessageId/

**Comment by thinhha** [comment]
Run Java PostCommit

**Comment by thinhha** [comment]
Run Java PostCommit

**Comment by thinhha** [comment]
Run Java PreCommit

**Comment by thinhha** [comment]
Run Java PostCommit

**Comment by lukecwik** [comment]
Run Java PostCommit

**Comment by thinhha** [comment]
Run Java PostCommit

**Comment by thinhha** [comment]
Run Java PostCommit

**Comment by thinhha** [comment]
Run Java PostCommit

**Comment by thinhha** [comment]
Run Java PostCommit

**Comment by thinhha** [comment]
Run Java PostCommit

**Comment by thinhha** [comment]
Run Java PostCommit

**Comment by thinhha** [comment]
Run Java PreCommit

**Comment by thinhha** [comment]
Run Java PreCommit

**Comment by thinhha** [comment]
Run Java PostCommit

**Comment by thinhha** [comment]
I added some logging to the IT to help with debug. Sorry for doing that here as I'm not sure how to run the IT locally.

**Comment by thinhha** [comment]
Run Java PostCommit

**Comment by thinhha** [comment]
Run Java PostCommit

**Comment by thinhha** [comment]
Hi @lukecwik, looks like the non-messageId coders are hard-coded in `PubsubUnboundedSource`:

Line 1088 of `beam/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSource.java`:
```
    @Override
    public Coder<PubsubMessage> getOutputCoder() {
      return outer.getNeedsAttributes()
          ? PubsubMessageWithAttributesCoder.of()
          : PubsubMessagePayloadOnlyCoder.of();
    }
```

This means that while the PubsubMessage is created with the messageId field using `PubsubIO.readMessagesWithMessageId()`/`PubsubIO.readMessagesWithAttributesAndMessageId()` , it does not show up in the subsequent step.

Looks like this is a required field for Autovalue. How do you think this case can be handled?

**Comment by lukecwik** [comment]
Add another constructor to PubsubUnboundedSource where you pass in a boolean as to whether you need the message id and update the getOuptutCoder to look at this property selecting between the 4 pubsub message coder variants. You'll also want to update https://github.com/apache/beam/blob/667bf8569460e959f02229a835f42fd4ad2ad9a4/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java#L1146 to look at this property and select between the coder with the message id and attributes or just the attributes coder based upon the new boolean that you added.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSource.java [hunk 6]
```diff
@@ -884,7 +884,7 @@ public PubsubMessage getCurrent() throws NoSuchElementException {
       if (current == null) {
         throw new NoSuchElementException();
       }
-      return new PubsubMessage(current.elementBytes, current.attributes);
+      return new PubsubMessage(current.elementBytes, current.attributes, current.recordId);
```
Comment (udim) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSource.java]: Should that be `current.messageId`?

Comment (udim) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSource.java]: nm, my mistake

**Comment by lukecwik** [comment]
It turns out that one of the Dataflow services doesn't output the message id which makes this change not work with Dataflow until the service is updated. @udim is working on getting that resolved.

**Comment by thinhha** [comment]
Thanks for the update @lukecwik! Good think the IT caught this.

I've added the changes we discussed above. We can rerun the postCommit tests once the dataflow backend change is complete.

**Comment by thinhha** [comment]
Run Java PostCommit

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubReadIT.java (hunk 5 above)
Comment (thinhha) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubReadIT.java]: Looks like the IT passed this time. Could you take another look?

Thanks!

**Comment by thinhha** [comment]
Run Java PostCommit

**Comment by thinhha** [comment]
All tests are passed. Friendly ping for review @lukecwik.

**Comment by lukecwik** [comment]
Thanks a lot for contributing this and dealing with the additional complexity that Dataflow added.

**Pull request merged**